Allow for the format of Routes of a Service to be configurable

duglin opened the pull request:
By default Routes are set to `ServiceName.Namespace.Domain`.
This change allows for someone to set a `RouteTemplate` field in the
`config-network` configMap with a golang template expression so they
can define any other URL pattern to use.

E.g. `{{.Name}}-{{.Namespace}}.{{.Domain}}` will use `-` instead of `.`
For us this is useful because we have an environment where the TLS cert
is wildcarded for `*.domain` and the wildcard only allows for one level
above the domain. Obviously, people need to be aware of what format they
choose because it could lead to restrictions. E.g., in this example, the use
of `-` in the Namespace could be problematic.  But this their choice.

As of now they can substitute Name, Namespace or Domain. We could
choose to add other values later if we wish by just adding them to the
data structure.

One thing we might want to consider is allowing the Route to be a URL
so that people can choose something other than just DNS names, for example:
`Domain/Namespace/Name` might be a valid choice for some environments.

Docs and test are coming but I wanted to get this out there for discussion
before I went too far in case the entire idea is rejected.

- - - - -

Another usecase for why this might be needed (see chat transcript below too), is that I may have a cluster (and I'm the sole owner of it) and I may choose to deploy various services into different namespaces. I'm using different namespaces not for multi-tenancy at a kube user level, but rather just to allow each service to create whatever resources it needs in their own namespace so I don't need to worry about conflicts at a kube resource level. E.g. each could create a secret of name "mySecrets" if it wanted w/o fear of interfering with other services. This could be important when I'm not really the author of those services and I'm just hosting them, and I don't really have control over the resources w/o forking things - which gets ugly.

So, in that case I may want to expose my services to the internet as `appName.clusterDomainName` w/o any `namespace` since the use of the namespace is strictly an internal hosting issue that my end-users will never care about, and therefore shouldn't be exposed to.

Signed-off-by: Doug Davis <dug@us.ibm.com>

**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```release-note
You can use the `config-network`'s `RouteTemplate` property to define for format
for your Routes. It will default to `{{.Name}}.{{.Namespace}}.{{.Domain}}`
```

duglin commented:
From a slack discussion @mattmoor and I had:

mattmoor [9:06 PM]
There's no good way (that I can think of) for it to gracefully degrade on collisions, and that also presents tenancy problems, if you hope to use namespaces for isolation

e.g. `foo-bar/baz` and `foo/bar-baz` always collide
and if one is serving traffic, there's no telling what creating the other would do
It there a reason to not just publish vanity aliases as their own VirtualService?

duglin [9:09 PM]
right, this isn't necessarily guaranteeing no-collision - rather it's about presenting an option for people who understand the limitations
That would require setting up more DNS entries, right?

mattmoor [9:10 PM]
No

duglin [9:10 PM]
elaborate on what you mean by "vanity aliases"

mattmoor [9:11 PM]
e.g. I have `mattmoor.io` pointed at a particular ingress gateway
foo.bar.mattmoor.io is allocated by Routes
but I can create an alias `www.mattmoor.io` that rewrites to `foo.bar.mattmoor.io`

duglin [9:11 PM]
right, someone needs to setup mattmoor.io to point to their cluster's ip

mattmoor [9:11 PM]
on the same gateway
that only requires a single DNS record for both

duglin [9:13 PM]
We have an env where \*.mycluster.myco.com is setup automatically as a DNS entry as part of the cluster 'mycluster'.  And for simple DNS resolution myapp.default.mycluster.myco.com works fine. But the tls cert is for *.mycluster.myco.com - which doesn't support myapp.default (edited) 
so, when I own the entire cluster. I'm ok with the name being myapp-default instead of myapp.default  - or even just myapp because it's my cluster
I'm not suggesting this would be something to set in a multi-tenant env - but either way, that's something for the admin to decide. Its not on a per app, or ns, basis.
The default is still what happens today.
Think of it this way.... it's my cluster and how I choose to divide things up internally (e.g. namespaces) is my impl/deployment choice - however, I may not choose to expose the namespace layer to my users. They don't care about that and as long a I name things appropriately I don't want them to see it either. So forcing me to include the namespace in the URL is an unnecessary thing - in this case. (edited) 

mattmoor [9:20 PM]
I could see us allowing it as a non-default option with big caveats in the ConfigMap

duglin [9:20 PM]
sure
the default is what happens today

mattmoor [9:20 PM]
obviously the `.svc.cluster.local` names via K8s Service wouldn't change
but I see this as somewhat similar to customizing that suffix

duglin [9:21 PM]
This is about letting people who know what they're doing, do what they want/need. But the default is the simple/guard-rails options.

mattmoor [9:21 PM]
probably worth discussing at the WG once Nghia is back

duglin [9:21 PM]
yep
I'll copy this chat into the PR for posterity  :slightly_smiling_face:

mattmoor [9:23 PM]
I think the other relevant element of this is that the deeper the domain you assign each cluster, the less likely a service like Let's Encrypt is to work because the rate limits are top-level

e.g. `foo.com` even if I've assigned `*.*.bar.baz.foo.com`
I'm also sure Evan will have thoughts
:wink:

duglin [9:25 PM]
that's actually some thing we discussed too. The idea of getting certs for the mutli-layers came up and the rate limiting issue was mentioned.

duglin commented:
FYI - I tried a test where I deployed a "helloworld" Ksvc to two different namespaces and set the RouteTemplate to `{{.Name}}.{{Domain}}`. As expected, each Ksvc was assigned the same Route value of `helloworld.example.com`. When I then tried to curl that URL I did get results successfully, but just from one of them - in this case the one in the `default` namespace. I don't know if that was just a random choice or there's code that will consistently pick the same one, but it seems to me that this is an acceptable outcome. The user defined an in-determinant configuration and as a result they got in-determinant results.

I then deleted the ksvc in the `default` namespace and the curl now hits the non-default-namespaced service. Which makes sense.

Re-creating the `default` version does not seem to change the `curl` results. It continues to go to the non-default-namespaced service.

Deleting the non-default one sends curl back to the default one.

This seems about right - a little bit of consistency within an inconsistent setup.

duglin commented:
ping @knative/serving-maintainers

bbrowning commented:
I have a similar need on the OpenShift side, where the convention is `name-namespace.domain` instead of `name.namespace.domain`. And wildcard certs are already setup for the former case so there's no special setup needed on behalf of the user.

markusthoemmes commented on pkg/network/network.go:
```diff
@@ -94,10 +103,16 @@ func NewConfigFromConfigMap(configMap *corev1.ConfigMap) (*Config, error) {
+
+	// If not in the config data then it'll be empty, which is ok, we'll
+	// just use the default value at that point
+	nc.RouteTemplate = configMap.Data[RouteTemplateKey]
```
Why not implement the defaulting behavior right here?

duglin replied:
You mean set it to name.ns.domain instead of "" ?
Because if someone sets the value in the configMap to "", or just erases it, then we'll still need to recover from that later on and I'd prefer to have the defaulting logic in only one spot.

markusthoemmes commented on pkg/reconciler/v1alpha1/route/route.go:
```diff
@@ -413,8 +423,57 @@ func objectRef(a accessor, gvk schema.GroupVersionKind) corev1.ObjectReference {
+	text := networkConfig.RouteTemplate
+
+	// Either this is the first time or the user-provided RouteTemplate changed
+	if savedTemplate == nil || text != savedText {
```
This caching behavior should be part of the config flow. It needs to become `DynamicConfig`. The current implementation is potentially racy.

duglin replied:
I agree that it's racy but I wasn't sure if we supported 2 controllers or not - I can fix that.
However, can you elaborate on what you mean by "part of the config flow"?
Do I need to use DynamicConfig if it's read-only in this func? Why doesn't "domainConfig" need it?

markusthoemmes replied:
Yes, presumably it will never surface as a real data race (however our controllers have a threadiness of 2 IIRC, so I might be wrong as well).

It is however a (to me) unexpected way of caching values. You can have a look at the autoscaler and how DynamicConfig is used there. Essentially, I'd like us to cache configuration related values in the configuration itself. Could we add the parsed template to the configuration?

markusthoemmes commented on pkg/reconciler/v1alpha1/route/route.go:
```diff
@@ -413,8 +423,57 @@ func objectRef(a accessor, gvk schema.GroupVersionKind) corev1.ObjectReference {
+	// These are the available properties they can choose from.
+	// We could add more over time - e.g. RevisionName if we thought that
+	// might be of interest to people.
+	data := struct {
```
Should the be a proper type vs. an anonymous one?

duglin replied:
I'm not a golang expert, does it matter if it's only used in this one func?

markusthoemmes replied:
It does not and it's not a golang thing. It's more a comment on code structure. If you have a defined type for this, you can just point at it in documentation and say: all fields of the type XYZ can be used in the gotemplate.

You can also reuse that type in tests potentially.

duglin replied:
Good reasons - done.  BTW, I switched "DomainTemplate" back to "RouteTemplate" :-)  I forgot that the template isn't defining the "domain", it's defining the entire "route" of the service. And I didn't want people to get confused and think that they're just defining the domain name instead of the entire URL's hostname value.

markusthoemmes commented on pkg/network/network.go:
```diff
@@ -45,6 +45,11 @@ const (
+	// RouteTemplateKey is the name of the configuration entry that
+	// specifies the golang template string to use to construct the
+	// Knative service's DNS name.
+	RouteTemplateKey = "routeTemplate"
```
Should this be "domainTemplate"?

duglin replied:
either works for me. I'll change it.

duglin replied:
Switched back to RouteTemplate since the value is more than just the domain, it's the full URL/route.

markusthoemmes replied:
I'd still say it's "domain". You're generating the (sub)-domain for the route. Point in case the function using this is called `routeDomain`

duglin replied:
But the string/URL generated has the top-most value, the function name, and we don't expect it to act as a domain for other things, do we?

duglin replied:
I can rename `routeDomain` if you want :-)  I think it's misleading

markusthoemmes commented on pkg/network/network.go:
```diff
@@ -56,6 +61,10 @@ type Config struct {
+
+	// RouteTemplate is the golang text template to use to generate the
+	// Route's URL for the Service.
+	RouteTemplate string
```
I'm wondering if we could/should make this part of the Route itself as well (and default to the configured value). That'd make it possible to throw whatever domain one might want in there.

duglin replied:
I think allowing a Route to have its own value is a good idea, but I'd still want to have a default global one that's configurable. We could do the Route-level one in a follow-on PR.

markusthoemmes commented on pkg/network/network.go:
```diff
@@ -94,10 +103,16 @@ func NewConfigFromConfigMap(configMap *corev1.ConfigMap) (*Config, error) {
+
+	// If not in the config data then it'll be empty, which is ok, we'll
+	// just use the default value at that point
+	nc.DomainTemplate = configMap.Data[DomainTemplateKey]
```
In this case I meant: Should we set `nc.DomainTemplate` to `DefaultDomainTemplate` here if the config map is empty?

duglin replied:
Sure I can do that. Just wondering though... what's the policy on defaults w.r.t. upgrades? Let's say the default value for a field is "foo" today, and the user never sets a value. Then they upgrade Knative and a new default value of "bar" has been defined. I can see an argument for existing resources sticking with 'foo' and I can see an argument for changing them to 'bar' since the user never specified a preference and it might be odd to have old and new resources behave differently w.r.t. a field that the user has never touched - thus a redeploy of the resource would result in a change in behavior.  Has Knative been following a pattern on this one?

@mattmoor this seems related to the API talks yesterday where you proposed exposing defaults as a way for people to learn about the more advanced config options.

duglin commented:
@markusthoemmes see if this new version aligns with what you were thinking.

markusthoemmes commented on pkg/network/network.go:
```diff
@@ -56,6 +67,25 @@ type Config struct {
+	// RouteTemplateText is the golang text template used to generate the
+	// Route's URL for the Service. This is the raw text string provided
+	// in the Network ConfigMap.
+	RouteTemplateText string
```
Do we really need that?

duglin replied:
We use this value in the error message in https://github.com/knative/serving/pull/3306/files#diff-001ba74612c5f969aa98bfefb45fbff9R448 so that we can show the value/text that couldn't be processed.

markusthoemmes commented on pkg/network/network.go:
```diff
@@ -94,10 +124,31 @@ func NewConfigFromConfigMap(configMap *corev1.ConfigMap) (*Config, error) {
+	if text == "" {
+		// Blank makes no sense so use our default instead
+		text = DefaultRouteTemplateText
+	}
```
If we really want the text in the config, should the Default make it in there as well? (Currently it doesn't)

duglin replied:
if we have a reason for adding it I can, but I can't think of any use for it outside of this func.

duglin commented:
/test pull-knative-serving-unit-tests

tcnghia commented:
/test pull-knative-serving-upgrade-tests

duglin commented:
@markusthoemmes I added a few more tests if you want to make sure I did those correctly. Look in route_test.go

duglin commented:
@tcnghia any thoughts on this one? Was it discussed during today's call?

tcnghia commented:
@ZhiminXiang how would this interfact with your TLS work?

tcnghia commented:
/lgtm
/approve

This looks good to me.  I was thinking we may also include sub route in the same template, but we can certainly follow up later.

mattmoor commented:
/hold

Want to discuss this in light of ongoing v1beta1 considerations around Route + `name:`

duglin commented:
@mattmoor can you elaborate on the issues?

duglin commented:
/test pull-knative-serving-integration-tests
/test pull-knative-serving-unit-tests

tcnghia commented:
@duglin I believe @mattmoor is talking about also including sub route information (in case of name traffic split) also in the same template.

ZhiminXiang commented:
/lgtm

duglin commented:
@tcnghia would that mean allowing additional "variables" to be used within the template string?

tcnghia commented:
@mattmoor @duglin were you able to discuss this change offline / here?

evankanderson commented:
One additional issue that just came up in discussion today is that DNS segments MUST be <63 characters, which suggests that the formatted hostname needs some way to constrain the length limits of the various donating components of the name.

duglin commented:
@mattmoor given the recent talks about "Steve", what's the status of this one in your eyes?

mattmoor commented:
/hold cancel

I think @evankanderson consideration is one of the things an Operator needs to understand here.

We'll want similar configuration for how `name:` becomes Kevin names in the related issues, and this is something we should consider validating in our webhook.

bbrowning commented:
@duglin I noticed the Release Note text is wrong here - it says `RouteTemplate` key of the configmap but should be `domainTemplate`, right?

duglin commented:
@bbrowning good catch, I forgot that I renamed it during the course of the PR, I fixed that.
@samodell is it too late for the v0.5 release? If so should I open a PR against some docs?
